aten/src/ATen/core/jit_type.h

```diff
@@ -324,25 +324,119 @@ inline c10::optional<T> merge_primitive(
+
+// If we see `a + b + c`  and know that a, b, and c are the same size and have two dimensions (WxH), 
+// then we can generate a fused kernel for them. 
+// That fused kernel would likely have indexing math to handling both the W and H dimensions. 
+// However, if we knew the WxH dimensions were contiguous, 
+// we can pretend like we only have a single dimension, simplifying the indexing logic. 
+// This can be performed even if the dimensions are transposed, 
+// as long as a, b, and c are transposed in the same way. 
+// We’d like to have the compiler be able to do this dimensionality reduction, 
+// but simply knowing sizes is not enough. 
+// We can extend profiling to also record stride information. 
+// Rather than recording specific strides, 
+// we can simply order the strides from smallest to largest with `stride_indices`
+// A contiguity marker on the smallest stride (c0) indicates the stride is precisely 1, 
+// otherwise a contiguity marker means that $stride_n = size_{n-1}*stride_{n-1}$
+struct CAFFE2_API Stride {
+  Stride() {}
+  Stride(const c10::optional<size_t>& stride_index, const c10::optional<bool>& contiguous, const c10::optional<size_t>& stride)
+  : stride_index_(stride_index)
+  , contiguous_(contiguous)
+  , stride_(stride) {}
+
+
+  bool operator ==(const Stride &b) const { 
+    return stride_index_ == b.stride_index_ && 
+      contiguous_ == b.contiguous_ && 
+      stride_ == b.stride_; 
+  }
+
+  c10::optional<size_t> stride_index_;
+  c10::optional<bool> contiguous_;
+  c10::optional<size_t> stride_;
+};
+
+template <>
+inline c10::optional<Stride> merge_primitive(
+    const c10::optional<Stride>& a,
+    const c10::optional<Stride>& b) {
+
+  c10::optional<Stride> left = a;
+  c10::optional<Stride> right = b;
+  if (!left.has_value()) {
+    left = {Stride()};
+  }
+  if (!right.has_value()) {
+    right = {Stride()};
+  }
+
+  auto merged_index = merge_primitive(left->stride_index_, right->stride_index_);
+  auto merged_cont = merge_primitive(left->contiguous_, right->contiguous_);
+  auto merged_stride = merge_primitive(left->stride_, right->stride_);
+  auto r = Stride (merged_index, merged_cont, merged_stride);
+  // normalize
+  if (!r.stride_index_.has_value() && !r.contiguous_.has_value() && !r.stride_.has_value())
+  {
+    return c10::optional<Stride>{};
+  }
+
+  return r;  
+}
+
+struct CAFFE2_API ShapeSymbol {
+    
+    // needed for use in `std::map`
+    ShapeSymbol(): value_(-1) {}
+    // is this symbol a fixed/static dimension
+    bool is_static() const { return value_ >= 0; };  
+    bool operator ==(const ShapeSymbol &b) const { return value_ == b.value_; }
+    bool operator <(const ShapeSymbol &b) const { return value_ < b.value_; }
+
+    static ShapeSymbol fromStaticSize(int64_t val) {return ShapeSymbol(val); }
+    int64_t static_size() const {
+      TORCH_CHECK(is_static());
+      return value_;
+    };
+
+    static ShapeSymbol newSymbol () {return -(++num_symbols); };
```

Comment:
First time this is called, it will return `-1` which is the same as `ShapeSymbol()`. Since `-1` is being used to encode a non-present shape, this seems wrong.

Reply:
`num_symbols` starts with 1 (i.e. there's always one unknown/undefined symbol). `newSymbol` increments first then returns, so the first new symbol should have the index of -2. I also ran a quick test to confirm and I'm indeed seeing `-2`